spec/primary_names_spec.lua

```diff
@@ -156,6 +211,60 @@ describe("Primary Names", function()
 				assert.are.equal(200000, _G.Balances[ao.id])
 			end
 		)
+
+		local validPrimaryUndernameNames = {
+			"1_test",
+			"1234_test",
+			"fsakdjhflkasjdhflkaf_test",
+			("").rep("1", 61) .. "_t",
+			"a_" .. ("").rep("1", 51),
+		}
+
+		for _, primaryName in ipairs(validPrimaryUndernameNames) do
+			it(
+				"should create a primary name request with undername and transfer the cost from the initiator to the protocol balance",
```

Comment:
It's not clear to me that unit testing this function over an array of valid undernames is testing for the right assertions, or at least not doing so in a a straightforward way. I think this makes it harder to evaluate your added code for correctness.

Reply:
Said differently, I'd have preferred to see unit tests specifically for, e.g.:
- utils.baseNameForName
- arns.getTokenCost
- assertValidUndername

and also vectorizing the happy path tests for affected primary names functions with their undername counterparts.

Reply:
Alternatively, I'd take a few integration test updates.

Reply:
Your integration test coverage probably exercises enough of what we need. Would hurt to pick off some of the low hanging fruit from the above list though.

Reply:
[13595b5](https://github.com/ar-io/ar-io-network-process/pull/392/commits/13595b58a4d30cddcb54fc48a39090301b2d42fa)